xds/internal/balancer/ringhash/e2e/ringhash_balancer_test.go

```diff
@@ -134,11 +142,336 @@ func (s) TestRingHash_ReconnectToMoveOutOfTransientFailure(t *testing.T) {
+
+	xdsServer, nodeID, _, xdsResolver, stop := e2e.SetupManagementServer(t, e2e.ManagementServerOptions{})
+	defer stop()
+
+	nonExistantServers := makeNonExistingBackends(t, 2)
+	servers, stop := startTestServiceBackends(t, 2)
+	defer stop()
+
+	newCluster1Name := "new_cluster_1"
+	newEdsService1Name := "new_eds_service_1"
+	newCluster2Name := "new_cluster_2"
+	newEdsService2Name := "new_eds_service_2"
+
+	ep1 := e2e.EndpointResourceWithOptions(e2e.EndpointOptions{
+		ClusterName: newEdsService1Name,
+		Localities: []e2e.LocalityOptions{{
+			Name:     "locality0",
+			Weight:   1,
+			Backends: backendOptions(t, nonExistantServers),
+		}},
+	})
+	ep2 := e2e.EndpointResourceWithOptions(e2e.EndpointOptions{
+		ClusterName: newEdsService2Name,
+		Localities: []e2e.LocalityOptions{{
+			Name:     "locality0",
+			Weight:   1,
+			Backends: backendOptions(t, servers),
+		}},
+	})
+	cluster1 := e2e.ClusterResourceWithOptions(e2e.ClusterOptions{
+		ClusterName: newCluster1Name,
+		ServiceName: newEdsService1Name,
+		Policy:      e2e.LoadBalancingPolicyRingHash,
+	})
+	cluster2 := e2e.ClusterResourceWithOptions(e2e.ClusterOptions{
+		ClusterName: newCluster2Name,
+		ServiceName: newEdsService2Name,
+		Policy:      e2e.LoadBalancingPolicyRingHash,
+	})
+	t.Log(cluster1, cluster2, ep1, ep2) // XXX
+	cluster := e2e.ClusterResourceWithOptions(e2e.ClusterOptions{
+		Type:        e2e.ClusterTypeAggregate,
+		ClusterName: "envoy.clusters.aggregate",
+		ChildNames:  []string{newCluster1Name, newCluster2Name},
+	})
+	route := e2e.DefaultRouteConfig("new_route", "test.server", newCluster1Name)
+	hashPolicy := v3routepb.RouteAction_HashPolicy{
+		PolicySpecifier: &v3routepb.RouteAction_HashPolicy_FilterState_{
+			FilterState: &v3routepb.RouteAction_HashPolicy_FilterState{
+				Key: "io.grpc.channel_id",
+			},
+		},
+	}
+	action := route.VirtualHosts[0].Routes[0].Action.(*v3routepb.Route_Route)
+	action.Route.HashPolicy = []*v3routepb.RouteAction_HashPolicy{&hashPolicy}
+	action.Route.ClusterSpecifier = &v3routepb.RouteAction_Cluster{Cluster: "envoy.clusters.aggregate"}
+	listener := e2e.DefaultClientListener("test.server", route.Name)
+
+	err := xdsServer.Update(context.Background(), e2e.UpdateOptions{
+		NodeID:    nodeID,
+		Clusters:  []*v3clusterpb.Cluster{cluster},
+		Routes:    []*v3routepb.RouteConfiguration{route},
+		Listeners: []*v3listenerpb.Listener{listener},
+	})
+	if err != nil {
+		t.Fatalf("failed to update server: %v", err)
+	}
+
+	conn, err := grpc.NewClient("xds:///test.server", grpc.WithResolvers(xdsResolver), grpc.WithTransportCredentials(insecure.NewCredentials()))
+	if err != nil {
+		t.Fatalf("failed to create client: %s", err)
+	}
+	conn.Connect() // force triggering resource fetch. This isn't great, see comment below.
+	defer conn.Close()
+	client := testgrpc.NewTestServiceClient(conn)
+
+	// Wait to give time to the LB policy to request the child clusters.
```

Comment:
If we can get the server to respond to a subset of resources using this: https://github.com/grpc/grpc-go/blob/master/internal/testutils/xds/e2e/server.go#L82, can we rely on channel state or some other state instead of sleeping?

Reply:
I think allowing resource subset works fine for this case actually, and we don't need any time-sensitive code. I've updated the code to use it.